scipy/sparse/tests/test_base.py

```diff
@@ -4132,44 +4193,55 @@ def test_scalar_idx_dtype(self):
     def test_binop_explicit_zeros(self):
         # Check that binary ops don't introduce spurious explicit zeros.
         # See gh-9619 for context.
-        a = csr_matrix([[0, 1, 0]])
-        b = csr_matrix([[1, 1, 0]])
+        a = self.csr_object([[0, 1, 0]])
+        b = self.csr_object([[1, 1, 0]])
         assert (a + b).nnz == 2
         assert a.multiply(b).nnz == 1
 
 
 TestCSR.init_class()
 
 
+class TestCSRMatrix(_MatrixMixin, TestCSR):
+    @classmethod
+    def spcreator(cls, *args, **kwargs):
+        with suppress_warnings() as sup:
+            sup.filter(SparseEfficiencyWarning, "Changing the sparsity structure")
+            return csr_matrix(*args, **kwargs)
+
+
+TestCSRMatrix.init_class()
```

Comment:
A bit confused as to how this initializer works without datsp having already been initialized; but, I suppose it works :shrug:

Reply:
The `init_class` method (about line 270) adds the `datsp` attribute to the **class** using the `spcreator` method to construct `datsp`. I don't think it is a normal idiom to use a classmethod to update the class after it has been defined. But it is how `test_base.py` has been set up for a while.

[detail that might be confusing:] In this case it looks like `TestCSR.init_class` has already been called, so I guess there is already a `datsp` attribute via inheritance. But that gets ignored after running `init_class` here to use the matrix version of `spcreator` to set the `datsp` attribute on the new class.